kratos/tests/cpp_tests/sparse_matrices/test_sparse_graph_serial.cpp

```diff
@@ -0,0 +1,227 @@
+        {{28,15},1.0},{{12,28},1.0},{{12,38},1.0},{{12,15},1.0},{{38,28},1.0},{{38,12},1.0},{{38,38},1.0},{{38,15},1.0},{{15,28},1.0},{{15,12},1.0},
+        {{15,38},1.0},{{8,6},1.0},{{18,6},1.0},{{6,8},1.0},{{6,18},1.0},{{28,31},1.0},{{28,22},1.0},{{28,39},1.0},{{31,28},1.0},{{31,22},1.0},
+        {{31,39},1.0},{{22,28},1.0},{{22,31},1.0},{{22,39},1.0},{{39,28},1.0},{{39,31},1.0},{{39,22},1.0},{{1,28},1.0},{{1,13},1.0},{{1,7},1.0},
+        {{28,1},1.0},{{28,13},1.0},{{28,7},1.0},{{13,1},1.0},{{13,28},1.0},{{13,13},1.0},{{13,7},1.0},{{7,1},1.0},{{7,28},1.0},{{7,13},1.0},
+        {{17,10},1.0},{{17,36},1.0},{{17,7},1.0},{{10,17},1.0},{{10,36},1.0},{{10,7},1.0},{{36,17},1.0},{{36,10},1.0},{{7,17},1.0},{{7,10},1.0},
+        {{25,30},1.0},{{25,9},1.0},{{14,30},1.0},{{30,25},1.0},{{30,14},1.0},{{30,9},1.0},{{9,25},1.0},{{9,30},1.0},};
+
+    return AMap;
+}
+
+
+template< class TSparseGraphType>
+bool CheckGraph(
+        const TSparseGraphType& rAgraph,
+        const MatrixMapType& rReferenceGraph)
+{
+    //check that all entries in Agraph are also in reference_A_map
+    const auto& rgraph = rAgraph.GetGraph();
+    for(const auto& item : rgraph)
+    {
+        const auto I = item.first;
+        for(auto J : item.second )
+        {
+            if(rReferenceGraph.find({I,J}) == rReferenceGraph.end()) //implies it is not present
+                KRATOS_ERROR << "Entry " << I << "," << J << "not present in A graph"  << std::endl;
+        }
+    }
+
+    //check that all the entries of reference_A_map are also in Agraph
+    for(const auto& item : rReferenceGraph)
+    {
+        auto I = item.first.first;
+        auto J = item.first.second;
+        if(rgraph.at(I).find(J) == rgraph.at(I).end())
+            KRATOS_ERROR << "Entry " << I << "," << J << " is in the reference graph but not in Agraph"  << std::endl;
+    }
+
+    return true;
+}
+
+bool CheckCSRGraphArrays(
+    const vector<SparseGraph::IndexType>& rRowIndices,
+    const vector<SparseGraph::IndexType>& rColIndices,
+    const MatrixMapType& rReferenceGraph)
+{
+    auto N = rRowIndices.size()-1;
+    for (IndexType I = 0; I < N; ++I)
+    {
+        for (auto k = rRowIndices[I]; k < rRowIndices[I + 1]; ++k)
+        {
+            IndexType J = rColIndices[k];
+            if (rReferenceGraph.find({I, J}) == rReferenceGraph.end()) //implies it is not present
+                KRATOS_ERROR << "Entry " << I << "," << J << "not present in A graph" << std::endl;
+
+            //check that that cols are ordered
+            if(k-rRowIndices[I] > 0)
+                if(rColIndices[k-1]>rColIndices[k])
+                    KRATOS_ERROR << "columns are not ordered in csr" << std::endl;
+        }
+    }
+    return true;
+}
+
+// Basic Type
+KRATOS_TEST_CASE_IN_SUITE(GraphConstruction, KratosCoreFastSuite)
+{
+    const auto connectivities = ElementConnectivities();
+    auto reference_A_map = GetReferenceMatrixAsMap();
+
+    SparseGraph Agraph;
```

Comment:
NOTE: SparseGraph is deliberately not threadsafe.

Parallel construction can be done as in the second test